Update GNUmakefile

danielsz opened the pull request:
Fix undefined reference to `__stack_chk_fail' when building. 

Platform: `Linux P14S 6.6.10-arch1-1 #1 SMP PREEMPT_DYNAMIC Fri, 05 Jan 2024 16:20:41 +0000 x86_64 GNU/Linux`

Apologies if a PR is unwarranted.

danielsz committed:
Update GNUmakefile

Fix undefined reference to `__stack_chk_fail'

matheusmoreira commented:
Good catch! I never ran into this problem during my development.

I wonder if I can somehow implement `__stack_chk_fail` instead. That way I can provide support for the compiler's correctness checking tools instead of simply turning them off. I'm also looking into how to implement the address and undefined behavior sanitizers but this stuff doesn't seem to be documented anywhere. [Even stackoverflow][question] failed to provide an answer despite a sizable bounty.

If there's no way to integrate with the compiler's stack protector, I agree that if should be disabled due to its dependency on external symbols. In that case, I will merge this.

I see you've edited the default value of the `CFLAGS` variable. Those defaults are meant to be overridden by the user. This flag is an essential option that impacts code generation in a way that could break the build if it's removed. So I recommend forcing it off by appending the option to `CFLAGS` later.

I suggest the following changes to your pull request:

https://github.com/lone-lang/lone/blob/fca96b0f5300ba1b1a42e1f32668f3d18975df0e/GNUmakefile#L92-L94

```diff
 # Disable strict aliasing and assume two's complement integers
 # even if CFLAGS contains options that affect this such as -O3
 CFLAGS += -fno-strict-aliasing -fwrapv

+# Disable the stack protector due to external symbol refefences
+CFLAGS += -fno-stack-protector
```

Please make the change so that you receive proper credits in case it's merged. Thank you for your contribution!

[question]: https://stackoverflow.com/q/77688456

matheusmoreira reviewed (changes requested):
I see you've edited the default value of the `CFLAGS` variable. Those defaults are meant to be overridden by the user. This flag is an essential option that impacts code generation in a way that could break the build if it's removed. So I recommend forcing it off by appending the option to `CFLAGS` later.

I suggest leaving the default `CFLAGS` alone and following the same approach as the strict aliasing and 2's complement code. Like this:

https://github.com/lone-lang/lone/blob/fca96b0f5300ba1b1a42e1f32668f3d18975df0e/GNUmakefile#L92-L94

```diff
 # Disable strict aliasing and assume two's complement integers
 # even if CFLAGS contains options that affect this such as -O3
 CFLAGS += -fno-strict-aliasing -fwrapv
+
+# Disable the stack protector due to external symbol refefences
+CFLAGS += -fno-stack-protector
```

matheusmoreira commented:
GCC stack smashing protection references:

 - [GCC Internals]
 - [OSDev Wiki]

When stack smashing protection is turned on, the compiler generates some fairly simple code:

```C
extern uintptr_t __stack_chk_guard;
noreturn void __stack_chk_fail(void);

void foo(const char* str)
{
	uintptr_t canary = __stack_chk_guard;

	/* ... */

	if ( (canary = canary ^ __stack_chk_guard) != 0 )
		__stack_chk_fail();
}
```

It looks like all we have to do is initialize `__stack_chk_guard` to a random number and provide a `__stack_chk_fail` function which exits the program. Linux provides 16 random bytes to every process via the auxiliary vector and lone already uses those to initialize its hash functions. It should be possible to use those random bytes to initialize the stack canary as well.

Would you like to try implementing this feature?

[GCC Internals]: https://gcc.gnu.org/onlinedocs/gccint/Stack-Smashing-Protection.html
[OSDev Wiki]: https://wiki.osdev.org/Stack_Smashing_Protector

danielsz committed:
Force CFLAG option

 This flag is an essential option that impacts code generation in a
 way that could break the build if it's removed.

danielsz commented:
I appreciate the thoughtfulness behind you letting me make changes in order to preserve the credits. That is the sign of a kind maintainer. Thank you!

matheusmoreira reviewed (approved):
Great! This resolves all my concerns.

matheusmoreira commented:
> I made the narrow changes that you've asked

Just merged it. Thanks for your contribution!

> created a new issue for the broader stack smashing protection

Thank you. I've assigned myself to that issue and will be doing research on the matter to see if there's any nuances I need to consider before implementing it.

Once I get it working, this PR can be reverted since the symbols will no longer be undefined. Perhaps you'll do the honors.
